### PR TITLE
Change limit to expression.

### PR DESCRIPTION
**Pull request by Shylock-Hg**
For the limit push down in Loop

Subjob of #2602

### CODE REVIEW

File: src/graph/planner/plan/Query.h [hunk 1]
```diff
@@ -663,14 +719,20 @@ class Limit final : public SingleInputNode {
...
+  Limit(QueryContext* qctx, PlanNode* input, int64_t offset, Expression* count)
+      : SingleInputNode(qctx, Kind::kLimit, input) {
+    offset_ = offset;
     count_ = count;
   }
 
   void cloneMembers(const Limit&);
 
  private:
   int64_t offset_{-1};
```
Comment (czpmango) [inline src/graph/planner/plan/Query.h:734]: Need change `offset` into an expression?

Comment (Shylock-Hg) [reply src/graph/planner/plan/Query.h:734]: offset won't push down

File: src/graph/planner/plan/Query.h [hunk 2]
```diff
@@ -652,9 +683,34 @@ class Limit final : public SingleInputNode {
...
-  int64_t count() const { return count_; }
+  // Get constant count value
+  int64_t count() const {
+    if (count_ == nullptr) {
+      return -1;
+    }
+    DCHECK(ExpressionUtils::isEvaluableExpr(count_));
+    QueryExpressionContext ctx;
+    return count_->eval(ctx).getInt();
+  }
```
Comment (czpmango) [inline src/graph/planner/plan/Query.h:703]: Is this interface necessary?

File: src/graph/planner/plan/Query.h (hunk 1 above)
Comment (czpmango) [reply src/graph/planner/plan/Query.h:734]: It is better to ensure consistency in use.
```
go from ....| offset abs(1)+3 limit abs(3)+1
go from ....| limit abs(3)+1,abs(3)
```

Comment (Shylock-Hg) [reply src/graph/planner/plan/Query.h:734]: This is used for new syntax for `go ... limit [..]`

File: src/graph/planner/plan/Query.h (hunk 2 above)
Comment (Shylock-Hg) [reply src/graph/planner/plan/Query.h:703]: Comptible.

**Comment by Shylock-Hg** [comment]
Subjob of #2602

File: src/graph/optimizer/rule/LimitPushDownRule.cpp [hunk 3]
```diff
@@ -50,6 +50,10 @@ StatusOr<OptRule::TransformResult> LimitPushDownRule::transform(
   const auto proj = static_cast<const Project *>(projGroupNode->node());
   const auto gn = static_cast<const GetNeighbors *>(gnGroupNode->node());
 
+  DCHECK(graph::ExpressionUtils::isEvaluableExpr(limit->countExpr()));
```
Comment (yixinglu) [inline src/graph/optimizer/rule/LimitPushDownRule.cpp:53]: Don't use DCHECK to debug your code if it's the regular branch you need to handle.

**Review by yixinglu — COMMENTED** [review]
I don't know what this PR is for? could you give some description?

File: src/graph/optimizer/rule/LimitPushDownRule.cpp (hunk 3 above)
Comment (Shylock-Hg) [reply src/graph/optimizer/rule/LimitPushDownRule.cpp:53]: This won't happen in correct state.

**Comment by Shylock-Hg** [comment]
> I don't know what this PR is for? could you give some description?

Push down limit to GN in Loop for each step.

**Review by nevermore3 — DISMISSED** [review]

File: src/graph/optimizer/rule/LimitPushDownRule.cpp (hunk 3 above)
Comment (yixinglu) [reply src/graph/optimizer/rule/LimitPushDownRule.cpp:53]: So why did you do the same check in line 54?

Comment (Shylock-Hg) [reply src/graph/optimizer/rule/LimitPushDownRule.cpp:53]: Handle exception scenes

**Comment by Sophie-Xie** [comment]
Include in #2853

**Pull request closed without merging**